### ISSUE TITLE
Make it possible for developers to install unreleased versions

### ISSUE THREAD

**Issue by tilgovi** [apache/incubator-annotator#92]
It would be helpful for everyone following along with the bleeding edge here to be able to consume packages from npm. We could publish continually, we just need to make it clear that these are developer convenience builds and not link to them from anywhere that suggests they are releases.

**Comment by Treora** [comment]
I suppose we could continue using the `dev` tag that we once used? E.g. [`@annotator/dom`](https://www.npmjs.com/package/@annotator/dom) has a release `0.1.0-dev.9d100c0f`. As discussed in the last call, we could try set up a hook to automatically build and publish it on npm whenever the master branch has changed, while generating a version number from the commit number.

On the other hand, I started wondering if we even want every development version to be pushed to the npm central registry. We discussed before that ideally people could just depend on the repository directly, using [a Git URL as dependency](https://docs.npmjs.com/files/package.json#git-urls-as-dependencies); e.g. writing `npm install https://github.com/apache/incubator-annotator#master`. The only problem we found was that our repo contains multiple packages. But perhaps we could still try make this work by making the top-level package the meta-package discussed in #90?

If we want to publish pre-built packages, we could also create a branch (or branches, or a separate repo) that contains those, e.g. using [`publish-to-git`](https://github.com/Rush/publish-to-git/) to automate this. Or we can just publish the latest tarball on the website: e.g. `npm install https://annotator.apache.org/packages/dom/0.2.0-dev.tar.gz` would be nice? While it would be good to make our packages readily available in the npm *ecosystem*, this need not mean they have to be on the npm *central registry*.

**Comment by tilgovi** [comment]
> While it would be good to make our packages readily available in the npm ecosystem, this need not mean they have to be on the npm central registry.

That's well said. Let me change the issue title to better reflect the goal.

**Comment by tilgovi** [comment]
> But perhaps we could still try make this work by making the top-level package the meta-package discussed in #90?

That won't work because I think lerna and/or yarn enforces that the top-level package must be private and won't publish it.

It might be possible to work _without_ lerna or yarn workspaces and just configure typescript and babel to know how to resolve everything, but we'd have to automate running yarn install separately for each package, they would all have their own yarn.lock, there would be no hoisting, and we might have to add all the development dependencies to each package.

I think we are better off not trying to publish the root, if we can find any way make the packages consumable otherwise.

Maybe GitHub Actions exist for this.

**Comment by tilgovi** [comment]
We could maybe use the GitHub package registry, too, if that's at all easier to integrate for whatever reason.

**Comment by Treora** [comment]
Note from today’s call: supposedly yarn 2 should be able to install packages from within a monorepo, so we could start with accepting that nightly users have to use yarn 2, and hopefully npm and other clients follow. The remaining tasks would be to test if this indeed works in yarn 2, and if so add a sentence about it to the readme.

**Comment by vrish88** [comment]
Hey all! Sorry I haven't been able to make the subsequent meetings. 

I'd be interested in using some of these non-versioned releases. Are there any constraints around publishing a "latest" build? Would you be open to a pr that does it in GH actions?

**Comment by Treora** [comment]
> I'd be interested in using some of these non-versioned releases. Are there any constraints around publishing a "latest" build? Would you be open to a pr that does it in GH actions?

I was just looking yesterday at the status of yarn 2, to see if I could use that to directly depend on this git repository (as explained above, our monorepo/multi-package structure makes it impossible to depend on using yarn 1 or npm). I don’t know if @tilgovi has gotten further with this?

In any case, even though directly depending on the git repo seems more elegant, it might take a while for us to move to yarn 2 (or not?), and it would force people to use yarn 2 until (if) monorepo dependencies are supported more widely. So to be pragmatic I think we should just auto-publish on npm indeed. I’d tag it `dev` instead of `latest`, to be clear it is not the latest release.

PRs would probably be very welcome! It depends on @tilgovi who was looking into this, but I don’t think we had anything set up yet.

**Comment by tilgovi** [comment]
After a long adventure and much confusion, I finally resolved this and the packages are publishing. The major hurdle was just that lerna had undocumented issues with using automation tokens and I had to use the beta v2 publishing providers from Travis.

In the process, I started running tests only against a current Node.js. That seems fine for now. We have linting and tooling designed to manage the compatibility, but let's keep the CI setup simple and snappy until we have a stronger need to guarantee stability.
